fem/bilininteg_hcurl.cpp

```diff
@@ -1802,6 +1802,515 @@ void PAHcurlH1Apply3D(const int D1D,
+      {
+         for (int ex = 0; ex < c_dofs1D; ++ex)
+         {
+            for (int dx = 0; dx < c_dofs1D; ++dx)
+            {
+               const int local_index = c_dofs1D * o_dofs1D + ey*c_dofs1D + ex;
+               y(local_index, e) += B(ex, dx) * vw[dx][ey];
+            }
+         }
+      }
+   });
+
+}
+
+static void PAHcurlApplyGradientTranspose2D(
+   const int c_dofs1D, const int o_dofs1D, const int NE,
+   const Array<double> &_B, const Array<double> &_G,
+   const Vector &_x, Vector &_y)
+{
+   auto B = Reshape(_B.Read(), c_dofs1D, c_dofs1D);
+   auto G = Reshape(_G.Read(), o_dofs1D, c_dofs1D);
+
+   auto x = Reshape(_x.Read(), 2 * c_dofs1D * o_dofs1D, NE);
+   auto y = Reshape(_y.ReadWrite(), c_dofs1D, c_dofs1D, NE);
+
+   MFEM_FORALL(e, NE,
+   {
+      double hw[c_dofs1D][o_dofs1D];
+      double vw[c_dofs1D][c_dofs1D];
+
+      // horizontal part (open x, closed y)
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int ex = 0; ex < o_dofs1D; ++ex)
+         {
+            hw[dy][ex] = 0.0;
+            for (int ey = 0; ey < c_dofs1D; ++ey)
+            {
+               const int local_index = ey*o_dofs1D + ex;
+               hw[dy][ex] += B(ey, dy) * x(local_index, e);
+            }
+         }
+      }
+
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int dx = 0; dx < c_dofs1D; ++dx)
+         {
+            for (int ex = 0; ex < o_dofs1D; ++ex)
+            {
+               y(dx, dy, e) += G(ex, dx) * hw[dy][ex];
+            }
+         }
+      }
+
+      // vertical part (open y, closed x)
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int ex = 0; ex < c_dofs1D; ++ex)
+         {
+            vw[dy][ex] = 0.0;
+            for (int ey = 0; ey < o_dofs1D; ++ey)
+            {
+               const int local_index = c_dofs1D * o_dofs1D + ey*c_dofs1D + ex;
+               vw[dy][ex] += G(ey, dy) * x(local_index, e);
+            }
+         }
+      }
+
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int dx = 0; dx < c_dofs1D; ++dx)
+         {
+            for (int ex = 0; ex < c_dofs1D; ++ex)
+            {
+               y(dx, dy, e) += B(ex, dx) * vw[dy][ex];
+            }
+         }
+      }
+   });
+}
+
+static void PAHcurlApplyGradient3D(const int c_dofs1D,
+                                   const int o_dofs1D,
+                                   const int NE,
+                                   const Array<double> &_B,
+                                   const Array<double> &_G,
+                                   const Vector &_x,
+                                   Vector &_y)
+{
+   auto B = Reshape(_B.Read(), c_dofs1D, c_dofs1D);
+   auto G = Reshape(_G.Read(), o_dofs1D, c_dofs1D);
+
+   auto x = Reshape(_x.Read(), c_dofs1D, c_dofs1D, c_dofs1D, NE);
+   auto y = Reshape(_y.ReadWrite(), (3 * c_dofs1D * c_dofs1D * o_dofs1D), NE);
+
+   MFEM_FORALL(e, NE,
+   {
+      double pxw1[c_dofs1D][c_dofs1D][c_dofs1D];
```

Comment:
Although these are "small" arrays, they can cause GPU stack memory failures for high order. Instead of 6 arrays, you can just use 2 arrays `w1` and `w2` of size `[c_dofs1D][c_dofs1D][c_dofs1D]`